### PR TITLE
fix(helm): fix regression with TLS flags/environment variables not being parsed

### LINKED ISSUE #4660: Helm v2.11.0-rc.3: not respecting --tls flags/environment variables
As discovered in #4651, there's a regression in v2.11.0-rc.3 that's not present in v2.10.0.

In v2.10.0, environment variables and flags are respected:

```
><> helm ls --tls --tls-cert=/foo
could not read x509 key pair (cert: "/foo", key: "/home/bacongobbler/.helm/key.pem"): can't load key pair from cert /foo and key /home/bacongobbler/.helm/key.pem: open /foo: no such file or directory
```

In v2.11.0-rc.3 though, these flags are not being respected:

```
><> ./linux-amd64/helm version -c
Client: &version.Version{SemVer:"v2.11.0-rc.3", GitCommit:"28d295be2a94115b786ee277dffcc2b5483bde47", GitTreeState:"clean"}
><> ./linux-amd64/helm ls --tls-cert=/foo --tls
could not read x509 key pair (cert: "/home/bacongobbler/.helm/cert.pem", key: "/home/bacongobbler/.helm/key.pem"): can't load key pair from cert /home/bacongobbler/.helm/cert.pem and key /home/bacongobbler/.helm/key.pem: open /home/bacongobbler/.helm/cert.pem: no such file or directory
```

### PR DESCRIPTION
**Pull request by bacongobbler**
This change fixes some of the assumptions made in #4590. The bug was triggered because environment variables were parsed well before execution (during `settings.AddFlagsTLS`), causing erroneous behaviour at runtime. By re-introducing `environment.Init()`, Helm can properly parse environment variables at the correct time.

One change that had to occur in this PR is the fact that we need to call `settings.Init()` each time we call `settings.AddFlagsTLS()`. This is because each command owns its own FlagSet, so we need to parse each flagset to read and propagate the environment variables correctly.

One other bug that was missed was the fact that we were maintaining two separate variables for each TLS value. Replacing the older code with the proper `settings` object makes the code much cleaner to read. Additionally, it fixes the issue where setting a flag or environment variable (e.g. `HELM_TLS_ENABLE=true` or `--tls`) would propagate to the `settings` object, but we'd be reading from `tlsEnable`, ignoring the flag/envvar.

A third, very odd behavioral bug was also missed dating back to when the TLS feature was first introduced. In prior versions of Helm, `helm init` and `addFlagsTLS` both [shared the same variables](https://github.com/helm/helm/blob/v2.10.0/cmd/helm/helm.go#L43-L48) for their respective TLS flags(`--tiller-tls` for `helm init`; `--tls` for everything else). Because they shared the same value, `helm init`'s default value [of the empty string](https://github.com/helm/helm/blob/v2.10.0/cmd/helm/init.go#L127) won over of `addFlagsTLS`'s [default value of `$HELM_HOME/ca.pem`](https://github.com/helm/helm/blob/v2.10.0/cmd/helm/helm.go#L317).

In practice, this code path caused the variable to be [expanded to the full path](https://github.com/helm/helm/blob/v2.10.0/cmd/helm/helm.go#L283-L285) of `$HELM_HOME/ca.pem` anyways, but it's why I had to add a check in this PR for `$HELM_HOME/ca.pem` in the root command's `PersistentPreRun`. If we find that `settings.TLSCaCertFile == "$HELM_HOME/ca.pem"`, we set it to home's default variable to preserve the old behaviour.

Fun fact: the "bug" can be recorded all the way back to [v2.3.0 when it was first released](https://github.com/helm/helm/blob/v2.3.0/cmd/helm/helm.go):

```
><> git diff
diff --git a/cmd/helm/helm.go b/cmd/helm/helm.go
index 4eee6795..3119015a 100644
--- a/cmd/helm/helm.go
+++ b/cmd/helm/helm.go
@@ -98,6 +98,7 @@ func newRootCmd(out io.Writer) *cobra.Command {
                Long:         globalUsage,
                SilenceUsage: true,
                PersistentPreRun: func(cmd *cobra.Command, args []string) {
+                       fmt.Println("TLS ca cert file:", tlsCaCertFile)
                        tlsCaCertFile = os.ExpandEnv(tlsCaCertFile)
                        tlsCertFile = os.ExpandEnv(tlsCertFile)
                        tlsKeyFile = os.ExpandEnv(tlsKeyFile)
><> make
><> ./bin/helm version -c
TLS ca cert file: 
Client: &version.Version{SemVer:"v2.3.0", GitCommit:"d83c245fc324117885ed83afc90ac74afed271b4", GitTreeState:"dirty"}
```

I've also added some unit tests that were originally used to verify the work being done in #4651 to ensure this regression doesn't occur again.

closes #4660 

Signed-off-by: Matthew Fisher <matt.fisher@microsoft.com>

### CODE REVIEW

**Comment by bacongobbler** [comment]
Ugh, after some manual testing this fix still doesn't appear to be working:

```
><> ./bin/helm version --tls --tls-cert=/foo
Client: &version.Version{SemVer:"v2.10+unreleased", GitCommit:"a0a30617dc6661642b2acb5e884632a33cefbfd4", GitTreeState:"clean"}
could not read x509 key pair (cert: "/home/bacongobbler/.helm/cert.pem", key: "/home/bacongobbler/.helm/key.pem"): can't load key pair from cert /home/bacongobbler/.helm/cert.pem and key /home/bacongobbler/.helm/key.pem: open /home/bacongobbler/.helm/cert.pem: no such file or directory
```

Marking this as a WIP for the time being.

**Comment by bacongobbler** [comment]
k, fixed

**Comment by bacongobbler** [comment]
(finally) fixed and working.

```
><> ./bin/helm version --tls
Client: &version.Version{SemVer:"v2.10+unreleased", GitCommit:"25de1d0eb9c7064980d4c9663c0516cbbc960cb4", GitTreeState:"clean"}
could not read x509 key pair (cert: "/home/bacongobbler/.helm/cert.pem", key: "/home/bacongobbler/.helm/key.pem"): can't load key pair from cert /home/bacongobbler/.helm/cert.pem and key /home/bacongobbler/.helm/key.pem: open /home/bacongobbler/.helm/cert.pem: no such file or directory
><> ./bin/helm version --tls --tls-cert=/foo
Client: &version.Version{SemVer:"v2.10+unreleased", GitCommit:"25de1d0eb9c7064980d4c9663c0516cbbc960cb4", GitTreeState:"clean"}
could not read x509 key pair (cert: "/foo", key: "/home/bacongobbler/.helm/key.pem"): can't load key pair from cert /foo and key /home/bacongobbler/.helm/key.pem: open /foo: no such file or directory
><> HELM_TLS_CERT=/foo ./bin/helm version --tls
Client: &version.Version{SemVer:"v2.10+unreleased", GitCommit:"25de1d0eb9c7064980d4c9663c0516cbbc960cb4", GitTreeState:"clean"}
could not read x509 key pair (cert: "/foo", key: "/home/bacongobbler/.helm/key.pem"): can't load key pair from cert /foo and key /home/bacongobbler/.helm/key.pem: open /foo: no such file or directory
```

File: cmd/helm/init.go [hunk 1]
```diff
@@ -166,6 +173,14 @@ func (i *initCmd) tlsOptions() error {
 				return errors.New("missing required TLS CA file")
 			}
 		}
+
+		// FIXME: refactor this all to pkg/helm/environment
```
Comment (michelleN) [inline cmd/helm/init.go]: Does the FIXME still need to be here?

**Review by michelleN — APPROVED** [review]

File: cmd/helm/init.go (hunk 1 above)
Comment (bacongobbler) [reply cmd/helm/init.go]: Yeah. I'd like to rip this all out in Helm 3 and replace it with the same flags used elsewhere. I'll update the comment to reflect that there's nothing immediately actionable here

**Comment by czka** [comment]
I'm confused. Is this supposed to be fixed in `2.11.0` or waiting for `2.11.1`?

My Helm `2.11.0` ignores `$HELM_TLS_CA_CERT`, `$HELM_TLS_CERT`, `$HELM_TLS_ENABLE`, `$HELM_TLS_KEY` and `$HELM_TLS_VERIFY` variables altogether. No matter how I set them. The only variable it honors is `$HELM_HOME`. When I put my pem files there, named as Helm expects them, I no longer have to specify `--tiller-tls-key`, `--tiller-tls-cert` and `--tls-ca-cert` on my command line (but I still have to set `--tls`!) for `helm ls --tls`, `helm version --tls` etc. Without `--tls` I keep getting 
`Error: cannot connect to Tiller` from `helm version` and `Error: transport is closing` from `helm ls`.

My `$HELM_HOME` is set to something else than `$HOME/.helm/` and my `$HOME/.helm/` does not even exist. I have my pem files in my `$HELM_HOME` named as Helm wants them by default (https://docs.helm.sh/helm). And, again, when my `$HELM_HOME` is exported my `helm` commands work fine when I supplement them with `--tls`.

I have one case when even `$HELM_HOME` is ignored though - `helm init` needs **all 3** `--tiller-tls-key`, `--tiller-tls-cert` and `--tls-ca-cert` paths set explicitly on the command line, always, also for `--upgrade`.

**Comment by czka** [comment]
In `2.11.0` `helm init` ignores `$HELM_TLS_CA_CERT`, `$HELM_TLS_CERT`, `$HELM_TLS_ENABLE`, `$HELM_TLS_KEY`, `$HELM_TLS_VERIFY` and even `$HELM_HOME` env variables. I have to set all the respective command line options explicitly for `helm init`. Is that as expected? Any subsequent `helm` command with same set of env variables does not require setting those options.

**Comment by czka** [comment]
Looks like `--tiller-tls-verify` setting is not being treated consistently.

Should the following really fail with `Error: missing required TLS key file`? The key file is right there in $HELM_HOME, as debug output says:

```
$ echo $HELM_HOME 
/home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm

$ echo $HELM_TLS_ENABLE 
true

$ helm init --dry-run --debug --service-account tiller --tiller-tls-verify
[debug] Host="", Key="/home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/key.pem", Cert="/home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/cert.pem", CA="/home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/ca.pem"

Error: missing required TLS key file
```
Exact same setup will not fail if I set the respective option in an env var, but then the TLS setup seems ignored, as you can see in the `helm init` output below: there should be `data:ca.cert`, `data:tlc.cert` and `data:tls.key` fields in the YAML, as well as `tiller-secret` Secret. Clearly `helm init` ignores the `HELM_TLS_VERIFY` env var setting.

```
$ export HELM_TLS_VERIFY="true"

$ helm init --dry-run --debug --service-account tiller
[debug] Host="", Key="/home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/key.pem", Cert="/home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/cert.pem", CA="/home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/ca.pem"

---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-deploy
  namespace: default
spec:
  replicas: 1
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: helm
        name: tiller
    spec:
      automountServiceAccountToken: true
      containers:
      - env:
        - name: TILLER_NAMESPACE
          value: default
        - name: TILLER_HISTORY_MAX
          value: "0"
        image: gcr.io/kubernetes-helm/tiller:v2.11.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          httpGet:
            path: /liveness
            port: 44135
          initialDelaySeconds: 1
          timeoutSeconds: 1
        name: tiller
        ports:
        - containerPort: 44134
          name: tiller
        - containerPort: 44135
          name: http
        readinessProbe:
          httpGet:
            path: /readiness
            port: 44135
          initialDelaySeconds: 1
          timeoutSeconds: 1
        resources: {}
      serviceAccountName: tiller
status: {}

---
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-deploy
  namespace: default
spec:
  ports:
  - name: tiller
    port: 44134
    targetPort: tiller
  selector:
    app: helm
    name: tiller
  type: ClusterIP
status:
  loadBalancer: {}
...
```

Only setting the pem files paths explicitely on the command line seem to work:
```
$ helm init --debug --dry-run --service-account tiller --tiller-tls-verify --tiller-tls-key /home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/key.pem --tiller-tls-cert /home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/cert.pem --tls-ca-cert /home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/ca.pem

[debug] Host="", Key="/home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/key.pem", Cert="/home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/cert.pem", CA="/home/msieczka/egnyte/repos-delphi/delphi-kubernetes/toolset/settings/helm/ca.pem"

---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-deploy
  namespace: default
spec:
  replicas: 1
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: helm
        name: tiller
    spec:
      automountServiceAccountToken: true
      containers:
      - env:
        - name: TILLER_NAMESPACE
          value: default
        - name: TILLER_HISTORY_MAX
          value: "0"
        - name: TILLER_TLS_VERIFY
          value: "1"
        - name: TILLER_TLS_ENABLE
          value: "1"
        - name: TILLER_TLS_CERTS
          value: /etc/certs
        image: gcr.io/kubernetes-helm/tiller:v2.11.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          httpGet:
            path: /liveness
            port: 44135
          initialDelaySeconds: 1
          timeoutSeconds: 1
        name: tiller
        ports:
        - containerPort: 44134
          name: tiller
        - containerPort: 44135
          name: http
        readinessProbe:
          httpGet:
            path: /readiness
            port: 44135
          initialDelaySeconds: 1
          timeoutSeconds: 1
        resources: {}
        volumeMounts:
        - mountPath: /etc/certs
          name: tiller-certs
          readOnly: true
      serviceAccountName: tiller
      volumes:
      - name: tiller-certs
        secret:
          secretName: tiller-secret
status: {}

---
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-deploy
  namespace: default
spec:
  ports:
  - name: tiller
    port: 44134
    targetPort: tiller
  selector:
    app: helm
    name: tiller
  type: ClusterIP
status:
  loadBalancer: {}

---
apiVersion: v1
data:
  ca.crt: <security ;)>
  tls.crt: <security ;)>
  tls.key: <security ;)>
kind: Secret
metadata:
  creationTimestamp: null
  labels:
    app: helm
    name: tiller
  name: tiller-secret
  namespace: default
type: Opaque

...
```

**Comment by czka** [comment]
Here's the shortest possible command line that installs Helm with TLS enabled proper, as of `2.11.0`, even if you have all the `HELM_HOME`, `HELM_TLS_ENABLE`, `HELM_TLS_VERIFY`, `HELM_TLS_KEY`, `HELM_TLS_CERT`, `HELM_TLS_CA_CERT` set as they should be:

`helm init --service-account tiller --tiller-tls-verify --tiller-tls-key $HELM_TLS_KEY --tiller-tls-cert $HELM_TLS_CERT --tls-ca-cert $HELM_TLS_CA_CERT`

**Comment by bacongobbler** [comment]
can you file a new bug please? Hard to track issues on closed/merged PRs. Thanks!

**Comment by czka** [comment]
@bacongobbler https://github.com/helm/helm/issues/4755. I'm removing my comments here.

**Pull request merged**